src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -141,589 +140,609 @@ void TemporalMemory::initialize(
-  set<Cell> _winnerCells;
-  set<UInt> _predictedColumns;
-  set<Cell> _predictedInactiveCells;
+    const Iterator& operator++()
+    {
+      NTA_ASSERT(!finished_);
+      calculateNext_();
+      return *this;
+    }
 
-  for (Cell cell : prevPredictiveCells)
-  {
-    UInt column = columnForCell(cell);
+  private:
 
-    if (activeColumns.find(column) != activeColumns.end())
+    UInt columnOf_(const SegmentOverlap& segmentOverlap) const
     {
-      _activeCells.insert(cell);
-      _winnerCells.insert(cell);
-      _predictedColumns.insert(column);
+      return segmentOverlap.segment.cell.idx / cellsPerColumn_;
     }
-  }
 
-  if (predictedSegmentDecrement_ > 0.0)
-  {
-    for (Cell cell : prevMatchingCells)
+    void calculateNext_()
     {
-      UInt column = columnForCell(cell);
-
-      if (activeColumns.find(column) == activeColumns.end())
+      if (activeColumn_ != activeColumnsEnd_ ||
+          activeSegment_ != activeSegmentsEnd_ ||
+          matchingSegment_ != matchingSegmentsEnd_)
+      {
+        current_.column = UINT_MAX;
+
+        if (activeSegment_ != activeSegmentsEnd_)
+        {
+          current_.column = std::min(current_.column,
+                                     columnOf_(*activeSegment_));
+        }
+
+        if (matchingSegment_ != matchingSegmentsEnd_)
+        {
+          current_.column = std::min(current_.column,
+                                     columnOf_(*matchingSegment_));
+        }
+
+        if (activeColumn_ != activeColumnsEnd_ &&
+            *activeColumn_ <= current_.column)
+        {
+          current_.column = *activeColumn_;
+          current_.isActiveColumn = true;
+          activeColumn_++;
+        }
+        else
+        {
+          current_.isActiveColumn = false;
+        }
+
+        current_.activeSegmentsBegin = activeSegment_;
+        while (activeSegment_ != activeSegmentsEnd_ &&
+               columnOf_(*activeSegment_) == current_.column)
+        {
+          activeSegment_++;
+        }
+        current_.activeSegmentsEnd = activeSegment_;
+
+        current_.matchingSegmentsBegin = matchingSegment_;
+        while (matchingSegment_ != matchingSegmentsEnd_ &&
+               columnOf_(*matchingSegment_) == current_.column)
+        {
+          matchingSegment_++;
+        }
+        current_.matchingSegmentsEnd = matchingSegment_;
+      }
+      else
       {
-        _predictedInactiveCells.insert(cell);
+        finished_ = true;
       }
     }
-  }
 
-  return make_tuple(_activeCells, _winnerCells,
-    _predictedColumns, _predictedInactiveCells);
-}
+    vector<UInt>::const_iterator activeColumn_;
+    vector<UInt>::const_iterator activeColumnsEnd_;
+    vector<SegmentOverlap>::const_iterator activeSegment_;
+    vector<SegmentOverlap>::const_iterator activeSegmentsEnd_;
+    vector<SegmentOverlap>::const_iterator matchingSegment_;
+    vector<SegmentOverlap>::const_iterator matchingSegmentsEnd_;
+    const UInt cellsPerColumn_;
```

Comment:
Actually it looks like this may not ever be serialized so may be fine. May not hurt to get in the habit of using `UInt32` explicitly though.